### DISCUSSION TITLE
Question : Conformer RNNT finetuning

### DISCUSSION THREAD

**Discussion by amejri** [NVIDIA-NeMo/NeMo#3725] [Q&A] [answered]
Hi guys, 

I am trying to finetune STT En Conformer RNNT for spanish. the problem is the following : when I try to change the tokenizer, the model is just such as training from scratch and it takes too much time for converging ( in reality it doesn't converge :p ). Does any one have an explanation, an idea or a trick to bypass this problem.

With best regards,

**Comment by nycsv** [comment, marked as answer]
Hi @amejri,

In case of changing the tokenizer, especially, the number of BPE, we should exclude some weights of part of the network when we load weights of pretrained model. For example, I'm finetuning the conformer-transducer model by adding the following contents to the yaml file like this:

NeMo/examples/asr/conf/conformer/conformer_transducer_bpe.yaml
```
name: "Conformer-Transducer-BPE"
init_from_nemo_model:
  model:
    path: /Users/eesungkim/pretrained_model/nemo_asr_set/stt_en_conformer_transducer_small.nemo 
    exclude: ["decoder.prediction.embed.weight","joint.joint_net.2.weight","joint.joint_net.2.bias"]
model:
  sample_rate: &sample_rate 16000
  compute_eval_loss: false 
  log_prediction: true 
  ctc_reduction: mean_batch

...
```

**Comment by VahidooX** [comment]
What is the effective batch you are using? How many epochs have you trained? What was your lr when you stopped the training?
If you effective batch size is less than 128 or 256, and your loss does not go down, you need to use grad accumulation or reduce your learning rate.

If you use a tokenizer with the same vocab size (1024), it is not necessary to exclude the decoder weights.

**Comment by titu1994** [comment]
@eesungkim gave a good suggestion, but we can further extend it. Say you know the vocab size of the pretrained model, and barring some special languages with large vocabularies, you can select a vocab size for the new language to match the old one - then you can simply load up the entire RNNT pretrained model - including the encoder, decoder and the joint all at once ! This would give you the fastest convergence out of the 3 options - don't use pretrained model (slowest), use pretrained encoder (moderate speed of convergence), use pretrained encoder decoder + joint (fastest to converge).

**Reply by nycsv** [reply]
@titu1994 Thanks for the helpful comments! There may be situations where it is necessary to train a model with a different vocab size depending on the language. For example, in the case of a Korean ASR, the performance is slightly different depending on whether it is modeled as charactor/syallable or vocab sizes. In addition, it would be better to use the partial load option when a smaller vocab size is needed to reduce the decoding speed.

**Reply by titu1994** [reply]
Yup agreed. Korean, Japanese Mandarin and a few other languages have such large base vocabularies that it overrides the pretrained checkpoint dimensions for it's vocab size c

Interestingly you mention decoding speed - for CTC models it's an implicit argmax so it's as fast as cuda can do argmax. 

For RNNT, Autoregressive decoding is costly but greedy decoding, especially with batched greedy decoding is around half of ctc speed now with larger batch sizez (there's been large speed boosts which will be in the 1.7 release, you can try out that branch as a pre-release). 

Though I guess for Beam Search, it definitely makes sense that the decoding is slow.
